met_connectors/src/frost/fetch.rs

```diff
@@ -71,50 +78,144 @@ fn parse_polygon(polygon: &Polygon) -> String {
     s
 }
 
-fn json_to_spatial_cache(resp: serde_json::Value, time: DateTime<Utc>) -> Result<DataCache, Error> {
-    let data = extract_data(resp, time)?;
+fn json_to_data_cache(
+    resp: serde_json::Value,
+    period: RelativeDuration,
+    num_leading_points: u8,
+    num_trailing_points: u8,
+    interval_start: DateTime<Utc>,
+    interval_end: DateTime<Utc>,
+) -> Result<DataCache, Error> {
+    let ts_vec = extract_data(resp, interval_start, period)?;
+
+    let processed_ts_vec = ts_vec
+        .into_iter()
+        .map(|((station_id, obses), location)| {
+            // TODO: preallocate?
+            // let ts_length = (end_time - first_obs_time) / period;
+            let mut data = Vec::new();
+
+            let mut curr_obs_time = interval_start - period * i32::from(num_leading_points);
+            let first_obs_time = obses
+                .first()
+                .ok_or(Error::MissingObs(
+                    "obs array from frost is empty".to_string(),
+                ))?
+                .time;
+
+            // handle misalignment of interval_start with ts, and leading missing values
+            if curr_obs_time != first_obs_time {
+                if first_obs_time < curr_obs_time {
+                    return Err(Error::Misalignment(
+                        "the first obs returned by frost is outside the time range".to_string(),
+                    ));
+                }
+
+                while first_obs_time >= curr_obs_time + period {
+                    data.push(None);
+                    curr_obs_time = curr_obs_time + period;
+                }
+
+                if first_obs_time != curr_obs_time + period {
+                    return Err(Error::Misalignment(
+                        "the first obs returned by frost is not aligned with the start time and period".to_string(),
+                    ));
+                }
+
+                curr_obs_time = first_obs_time;
+            }
+
+            // insert obses into data, with Nones for gaps in the series
+            for obs in obses {
+                if curr_obs_time == obs.time {
+                    data.push(Some(obs.body.value));
+                    curr_obs_time = curr_obs_time + period;
+                } else {
+                    while curr_obs_time < obs.time {
+                        data.push(None);
+                        curr_obs_time = curr_obs_time + period;
+                    }
+                    if curr_obs_time == obs.time {
+                        data.push(Some(obs.body.value));
+                        curr_obs_time = curr_obs_time + period;
+                    } else {
+                        return Err(Error::Misalignment(
+                            "obs misaligned with series".to_string(),
+                        ));
+                    }
+                }
```

Comment:
Do we need the first if condition? Isn't it already handled inside the while loop?

Reply:
Done